genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobTerminationServiceImpl.java

```diff
@@ -0,0 +1,120 @@
+import com.netflix.genie.web.services.JobSearchService;
+import io.grpc.stub.StreamObserver;
+import lombok.extern.slf4j.Slf4j;
+import net.devh.springboot.autoconfigure.grpc.server.GrpcService;
+
+import javax.validation.constraints.NotBlank;
+import java.util.Map;
+
+/**
+ * Extension of {@link JobTerminationServiceGrpc.JobTerminationServiceImplBase}
+ * to provide functionality for killing jobs running on agents.
+ * TODO Register with HeartBeatService to listen for agent stream to become inactive and clean it up.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@GrpcService(
+    value = JobTerminationServiceGrpc.class,
+    interceptors = {
+        SimpleLoggingInterceptor.class,
+    }
+)
+@Slf4j
+public class GRpcJobTerminationServiceImpl
+    extends JobTerminationServiceGrpc.JobTerminationServiceImplBase
+    implements JobKillServiceV4 {
+
+    private final Map<String, StreamObserver<JobKillListenerResponse>> parkedJobKillResponseObservers =
+        Maps.newConcurrentMap();
+    private final JobSearchService jobSearchService;
+
+    /**
+     * Constructor.
+     *
+     * @param jobSearchService Job search service
+     */
+    public GRpcJobTerminationServiceImpl(final JobSearchService jobSearchService) {
+        this.jobSearchService = jobSearchService;
+    }
+
+    /**
+     * Register to be notified when a kill request for the job is received.
+     *
+     * @param request          Request to register for getting notified when server gets a job kill request.
+     * @param responseObserver The response observer
+     */
+    @Override
+    public void registerForKillNotification(
+        final JobKillListenerRequest request,
+        final StreamObserver<JobKillListenerResponse> responseObserver
+    ) {
+        parkedJobKillResponseObservers.put(request.getJobId(), responseObserver);
+    }
+
+    @Override
+    public void killJob(final @NotBlank(message = "No job id entered. Unable to kill job.") String jobId,
+                        final @NotBlank(message = "No reason provided.") String reason) throws GenieException {
+
+        final StreamObserver<JobKillListenerResponse> responseObserver =
+            cleanUpAndGetParkedJobKillResponseObserver(jobId);
+
+        if (responseObserver != null && !jobSearchService.getJobStatus(jobId).isFinished()) {
+            responseObserver.onNext(
+                JobKillListenerResponse.newBuilder().build()
+            );
+            responseObserver.onCompleted();
+
+            log.info("Agent notified for killing the job ");
+        } else if (responseObserver == null) {
+            log.error("Job not killed. No response observer found for killing the job with id - {} ", jobId);
+            throw new GenieServerException(
+                "Job not killed. No response observer found for killing the job with id - {} " + jobId);
+        } else {
+            log.info("v4 job {} was already finished when the kill request came", jobId);
```

Comment:
Wouldn't this be more readable as:
```
if observer == null
  print something
if not finished
  print something
else
  send kill
```

Personal preference, but seems more clear/elegant to me than checking the same `null` twice